File: py/scene.py
from util.database.database import Database
from util.database.scene import fetch_scene, add_scene, update_scene, load_scene, save_scene
from util.database.scene_animation import fetch_scene_animations_from_scene, remove_scene_animations, add_scene_animation


class Scene:
    def __init__(self, id: int | None = None, name: str | None = None,
                 description: str | None = None, scene_animations: list = []):
        self.id = id
        self.scene_animations = scene_animations

        if name != None and description != None:
            self.name = name
            self.description = description

        elif self.id != None:
            scene_data = fetch_scene(self.id)

            if scene_data:  # Check if scene exists
                self.name = scene_data['name']
                self.description = scene_data['description']

        if len(self.scene_animations) == 0:  # Fetch animations to run
            scene_animations_data = fetch_scene_animations_from_scene(self.id)

            if scene_animations_data:
                self.scene_animations = scene_animations_data

    def sync_changes_to_db(self, new: bool = False) -> bool:
        if new:
            max_id = Database.fetchone_from_db('SELECT MAX(id) FROM scenes', {})[0]

            self.id = 0
            if max_id != None:
                self.id = max_id + 1

            return add_scene(self.id, self.name, self.description)

        return update_scene(self.id, self.name, self.description)

    def load(self) -> list | bool:
        """Load scene into workspace and return running animations.
        """
        if not load_scene(self.id):
            return False

        return self.scene_animations

    def save(self, running_animations: list) -> bool:
        """Save scene to database.
        """
        remove_scene_animations(self.id)
        
        for animation in running_animations:
            # 'scene_id': scene_animation_data[0],
            # 'animation_id': scene_animation_data[1],
            # 'section_id': scene_animation_data[2],
            # 'color_sequence_id': scene_animation_data[3]
            if animation.color_sequence == None:
                continue
            
            add_scene_animation(self.id, animation.id, animation.section_id, animation.color_sequence.id)
            
        if len(self.scene_animations) == 0:  # Fetch animations to run
            scene_animations_data = fetch_scene_animations_from_scene(self.id)

            if scene_animations_data:
                self.scene_animations = scene_animations_data
        
        return save_scene(self.id)

File: src/Scene.cpp
#include "Scene.hpp"

#include "util/database/SceneDB.hpp"
#include "util/database/SceneAnimationDB.hpp"
#include "util/database/Database.hpp"
#include "animation/Animation.hpp"

Scene::Scene(std::optional<int> id,
             std::optional<std::string> name,
             std::optional<std::string> description,
             std::optional<std::vector<std::map<std::string, int>>> scene_animations)
    : id(id.value_or(-1)), 
      scene_animations(scene_animations ? 
                       *scene_animations 
                       : std::vector<std::map<std::string, int>>())
{
    if (name.has_value() && description.has_value()) {
        this->name = name.value_or("");;
        this->description = description.value_or("");;
    } else if (id) {
        auto scene_data = SceneDB::fetch_scene(id.value());

        this->name = scene_data["name"];
        this->description = scene_data["description"];
    }

    if (this->scene_animations.empty() && this->id >= 0) {
        auto animations = SceneAnimationDB::fetch_scene_animations_from_scene(this->id);
        
        this->scene_animations = animations;
    }
}

bool Scene::sync_changes_to_db(bool is_new) {
    if (is_new) {
        auto response = Database::fetchone_from_db("SELECT MAX(id) FROM scenes;", {});
        
        if (!response.empty()) {
            int max_id = std::stoi(response[0]);

            if (max_id < 0) {
                max_id = 0;
            }
            this->id = max_id + 1;
        } else {
            this->id = 0; // Start from 0 if no scenes exist
        }

        return SceneDB::add_scene(this->id, this->name, this->description);
    }

    return SceneDB::update_scene(this->id, this->name, this->description);
}

std::optional<std::vector<std::map<std::string, int>>> Scene::load() {
    if (!SceneDB::load_scene(this->id)) {
        return std::nullopt;
    }

    return std::optional{this->scene_animations};
}

bool Scene::save(const std::vector<std::map<std::string, int>>& running_animations) {
    SceneAnimationDB::remove_scene_animations(this->id);

    for (const auto& animation : running_animations) {
        SceneAnimationDB::add_scene_animation(
            this->id,
            animation.at("id"),
            animation.at("section_id"),
            animation.at("color_sequence_id")
        );
    }

    if (this->scene_animations.empty()) {
        auto animations = SceneAnimationDB::fetch_scene_animations_from_scene(this->id);
        this->scene_animations = animations;
    }

    return SceneDB::save_scene(this->id);
}
